Add paged tables to render data

javierluraschi opened the pull request:
Almost there, need to address the SQL comment from JJ...

<img width="946" alt="screen shot 2016-07-26 at 2 42 00 pm" src="https://cloud.githubusercontent.com/assets/3478847/17156550/479636da-533f-11e6-8368-8e8a3051486d.png">

<img width="934" alt="screen shot 2016-07-26 at 3 42 44 pm" src="https://cloud.githubusercontent.com/assets/3478847/17158118/cdf7ff98-5348-11e6-812f-2e75c60e9b56.png">

javierluraschi commented:
@jjallaire:

1. Removed `jquery` dependencies. Seems some of the functionality was used to generate TOC, but I rendered one doc with it and is looking good. Let me know if there is something else I should check.

2. Ran notebook with SQL chunk. It's currently rendering as text, is this what you would expect from the comment you mention about the table coming from `kable`? Would be nice to render those as pageable tables as well, I can take a look if we are interested in this. The SQL chunks looks as follows (unchanged to my understanding):

Rendering SQL chunk from UI:
<img width="925" alt="screen shot 2016-07-26 at 4 12 30 pm" src="https://cloud.githubusercontent.com/assets/3478847/17158596/54282dc4-534c-11e6-9671-e674e0cb92b5.png">

Rendering SQL chunk with `rmarkdown::render("main.Rmd", "html_notebook")`:
<img width="927" alt="screen shot 2016-07-26 at 4 12 58 pm" src="https://cloud.githubusercontent.com/assets/3478847/17158597/542974f4-534c-11e6-9aa8-fd1477c73e2f.png">

javierluraschi commented:
@jmcphers planning to check-in the pageable-table here and then integrate into the IDE the `pageable.js` and `pageable.css` files. However, might be easier to iterate here if you have design feedback.

Here is a live demo BTW, functionality is pretty basic at the moment: https://s3-us-west-2.amazonaws.com/jluraschi-temp/rmarkdown/main.nb.html - I could consider scrolling animations if that sounds interesting.

jjallaire commented:
Some UX comments:

1. In `kable` character columns are left-aligned by default. I'd repeat that behavior here.

2. I think we need to display the total number of records somewhere (and which "window" we are currently in). I also think that we should have the 1 2 3 >> type UI treatment for paging. e.g. something like we see here in DataTables:

<img width="824" alt="screen shot 2016-07-27 at 8 22 31 am" src="https://cloud.githubusercontent.com/assets/104391/17175035/530a5e58-53d3-11e6-8fd0-b0aeb1d3c35b.png">

jjallaire commented on DESCRIPTION:
```diff
@@ -13,12 +13,6 @@ Authors@R: c(
   person("Aron", "Atkins", role = "aut", email = "aron@rstudio.com"),
   person("Rob", "Hyndman", role = "aut", email = "robjhyndman@gmail.com"),
   person(family = "RStudio, Inc.", role = "cph"),
-  person(family = "jQuery Foundation", role = "cph",
```
We are still using jquery just don't need the explicit paring with your JS file (i.e. it's already included elsewhere automatically). The reason we had it there for stickytableheaders was to ensure it was included in the correct order (i.e. before stickytableheaders).

jjallaire commented on R/html_document.R:
```diff
@@ -241,9 +241,7 @@ html_document <- function(toc = FALSE,
 
     # dependencies
     extra_dependencies <- append(extra_dependencies,
```
These all need to be restored.

javierluraschi replied:
Reverted the jquery removal commit.

jjallaire commented on R/html_document_base.R:
```diff
@@ -85,8 +85,7 @@ html_document_base <- function(smart = TRUE,
     format_deps <- list()
     format_deps <- append(format_deps, extra_dependencies)
     if (!is.null(theme)) {
-      format_deps <- append(format_deps, list(html_dependency_jquery(),
```
This needs to be restored.

jjallaire commented on R/knit_print.R:
```diff
@@ -1,11 +1,19 @@
     if (identical(context$df_print, knitr::kable)) {
-      res = paste(c('<div class="kable-table">', '', knitr::kable(x), '', '</div>'),
-                  collapse = "\n")
+      res = paged_table_html(x)
```
We can't do this here because this code runs for all rmarkdown formats (i.e. this would result in your paged_table_html running for PDFs!!!). Rather, for the `html_notebook` format we need to pass our df printing function for the `df_print` argument to `output_format` (i.e. where we currently pass the string "kable" in `html_notebook` we pass a function).

jjallaire commented on inst/rmd/h/jquery-AUTHORS.txt:
```diff
@@ -1,217 +0,0 @@
-Authors ordered by first contribution.
```
Restore jquery as mentioned before.

jjallaire commented on inst/rmd/h/jqueryui-1.11.4/index.html:
```diff
@@ -1,505 +0,0 @@
-<!doctype html>
```
Restore jqueryui as mentioned before.

jjallaire commented:
Note that when you do integrate into the IDE you can just read the .js and .css files via `system.file` (since we install an embedded version of rmarkdown we are guaranteed to always have the latest version).

javierluraschi commented:
@jjallaire Thanks, following changes performed:

1. Reverted removal of jquery change, scoped removal to `html_notebook.R`
2. Scoped use of `paged_table_html` to only the `html_notebook`
3. Left-aligned `character` and `factor` columns
4. Added navigation bar with page indexes and row-count info box

<img width="931" alt="screen shot 2016-07-27 at 1 28 29 pm" src="https://cloud.githubusercontent.com/assets/3478847/17191276/0fcd8ff4-53fe-11e6-9f6e-372a2a3ab4eb.png">

jjallaire commented:
Maybe capitalize "previous" and "next" ?

javierluraschi commented:
Sure!

<img width="935" alt="screen shot 2016-07-27 at 2 08 42 pm" src="https://cloud.githubusercontent.com/assets/3478847/17192513/b4feceac-5403-11e6-84f1-4286b4fa3783.png">

jjallaire commented:
We could implement the knit sql compatibility by allowing rmarkdown to set a knit sql render function (as a knit option, sort of like we do right here: https://github.com/rstudio/rmarkdown/blob/master/R/html_notebook.R#L128-L137). This option could be sql.print, e.g.

```r
knitr::opts_knit$set(sql.print = paged_table_html)
```

Then knitr would check for it here: https://github.com/yihui/knitr/blob/master/R/engine.R#L471

Alternatively, you could just look for `<div class="knitsql-table">` and automatically transform it to your structured/implementation. 

I don't have a strong opinion about which way is best.

@yihui

yihui commented:
I don't have a strong opinion, either.

jjallaire commented:
Providing a hook within knitr would allow for a more compact (JSON) representation of the data frame within the published document. @javierluraschi Let's put a PR into knitr to make the sql.print method pluggable and then use that to redirect to our data frame renderer.

javierluraschi commented:
@jjallaire @yihui I went for the `knitr` option following the option pattern from `sql.max.print`.

Here is the `knitr` PR: https://github.com/yihui/knitr/pull/1257

Sample output:

<img width="923" alt="screen shot 2016-07-28 at 12 18 54 am" src="https://cloud.githubusercontent.com/assets/3478847/17204192/e7cff5ea-5458-11e6-87df-b50a472c572f.png">

Sample notebook: https://beta.rstudioconnect.com/content/1369/

jjallaire commented:
A couple more suggested refinements:

1. Do you think we need a line that runs above the footer area? (hard to know until you actually see it I guess)

2. When there are less than 10 records I think we should: (a) Just display e.g. "3 records" (rather than "1-3 of 3"); (b) Get rid of the Previous/Next treatment entirely since it's a no-op.

javierluraschi commented:
I think it looks better, updated:

<img width="922" alt="screen shot 2016-07-28 at 8 57 06 am" src="https://cloud.githubusercontent.com/assets/3478847/17220028/91b70b22-54a2-11e6-8a9e-25348c09d1b1.png">

<img width="929" alt="screen shot 2016-07-28 at 8 57 11 am" src="https://cloud.githubusercontent.com/assets/3478847/17220029/9408c55a-54a2-11e6-8853-2e5c9419a60a.png">

jjallaire commented:
Agree it looks better. A couple other small things I noticed:

- Column headers should reflect the alignment of data

- Scratch the page number as well when Next and Previous are omitted
On Thu, Jul 28, 2016 at 12:06 PM Javier Luraschi <notifications@github.com>
wrote:

> I think it looks better, updated:
>
> [image: screen shot 2016-07-28 at 8 57 06 am]
> <https://cloud.githubusercontent.com/assets/3478847/17220028/91b70b22-54a2-11e6-8a9e-25348c09d1b1.png>
>
> [image: screen shot 2016-07-28 at 8 57 11 am]
> <https://cloud.githubusercontent.com/assets/3478847/17220029/9408c55a-54a2-11e6-8853-2e5c9419a60a.png>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/rmarkdown/pull/751#issuecomment-235942715>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXxzP_H2waQizAXvRKdHWTbtIJeTqXks5qaNObgaJpZM4JVrE8>
> .
>

javierluraschi commented:
Updated, fixed also 0-records case:

<img width="934" alt="screen shot 2016-07-28 at 9 27 02 am" src="https://cloud.githubusercontent.com/assets/3478847/17220764/9015e2ea-54a5-11e6-8a32-e00c244cda2c.png">

<img width="925" alt="screen shot 2016-07-28 at 9 27 16 am" src="https://cloud.githubusercontent.com/assets/3478847/17220775/98509838-54a5-11e6-913d-33c59b8a15af.png">

<img width="917" alt="screen shot 2016-07-28 at 9 27 25 am" src="https://cloud.githubusercontent.com/assets/3478847/17220776/9855dbb8-54a5-11e6-9941-79fe10752042.png">

jjallaire commented:
One thing we did in SQL printing within knitr is detect when the first column is an ID column and in that case left align it (as right aligning an ID column looks odd, the latest screenshot is a case in point). Here's the code we used for this:

https://github.com/yihui/knitr/blob/master/R/engine.R#L454-L457

I'm going to go ahead and merge this and we can make further refinements on master. This feature looks great!

hadley commented:
If this is going to replace the default tibble display, I'd really like to surface the variable types somehow (maybe in small grey text beneath the names)

jjallaire commented:
Yes, definitely! @javierluraschi <https://github.com/javierluraschi> let's
do this



On Fri, Jul 29, 2016 at 10:33 AM, Hadley Wickham <notifications@github.com>
wrote:

> If this is going to replace the default tibble display, I'd really like to
> surface the variable types somehow (maybe in small grey text beneath the
> names)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/rmarkdown/pull/751#issuecomment-236196512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXxxfAjQabB13IL2gcaCK5UaeLRhhOks5qag8sgaJpZM4JVrE8>
> .
>

javierluraschi commented:
Noted! I'll finish the ide integration and then perform another pass at the comments in this pr.

javierluraschi commented:
Following up with https://github.com/rstudio/rmarkdown/compare/feature/pagedtable-types?expand=1 for the couple remaining feedback points.
